sound/soc/sof/topology.c

```diff
@@ -1385,67 +1375,170 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
+	fir_data_p = private->data + SOF_EFFECT_PAYLOAD_OFFSET;
+
+	fir = kzalloc(ipc_size, GFP_KERNEL);
+	if (!fir)
 		return -ENOMEM;
 
-        /* configure IIR EQ IPC message */
-	eq->comp.hdr.size = sizeof(*eq);
-	eq->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
-	eq->comp.id = swidget->comp_id;
-	eq->comp.pipeline_id = index;
+	/* configure fir IPC message */
+	fir->comp.hdr.size = ipc_size;
+	fir->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
+	fir->comp.id = swidget->comp_id;
+	fir->comp.type = SOF_COMP_EQ_FIR;
+	fir->comp.pipeline_id = index;
 
-	eq->comp.type = SOF_COMP_EQ_IIR;
-	ret = sof_parse_tokens(scomp, eq, eq_iir_tokens,
-			       ARRAY_SIZE(eq_iir_tokens), private->array,
-			       le32_to_cpu(private->size));
-	if (ret) {
-		dev_err(sdev->dev, "error: parse EQ tokens failed %d\n",
-			private->size);
-		goto err;
-	}
+	memcpy(&fir->data, fir_data_p, fir_data_size);
 
-	ret = sof_parse_tokens(scomp, &eq->config, comp_tokens,
+	fir->size = fir_data_size;
+
+	ret = sof_parse_tokens(scomp, &fir->config, comp_tokens,
 			       ARRAY_SIZE(comp_tokens), private->array,
-			       le32_to_cpu(private->size));
-	if (ret) {
-		dev_err(sdev->dev, "error: parse EQ.cfg tokens failed %d\n",
+			       SOF_EFFECT_PAYLOAD_OFFSET);
+	if (ret != 0) {
+		dev_err(sdev->dev, "error: parse fir.cfg tokens failed %d\n",
 			le32_to_cpu(private->size));
 		goto err;
 	}
 
-	dev_dbg(sdev->dev, "eq iir %s created\n", swidget->widget->name);
+	sof_dbg_comp_config(scomp, &fir->config);
+
+	swidget->private = (void *)fir;
+
+	ret = sof_ipc_tx_message(sdev->ipc, fir->comp.hdr.cmd, fir,
+				 ipc_size, r, sizeof(*r));
+	if (ret >= 0)
+		return ret;
+err:
+	kfree(fir);
+	return ret;
+}
+
+static int sof_effect_iir_load(struct snd_soc_component *scomp, int index,
+			       struct snd_sof_widget *swidget,
+			       struct snd_soc_tplg_dapm_widget *tw,
+			       struct sof_ipc_comp_reply *r)
+
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_soc_tplg_private *private = &tw->priv;
+	struct sof_ipc_comp_eq_iir *iir;
+	size_t ipc_size = 0, iir_data_size = 0;
+	void *iir_data_p;
+	int ret;
+
+	iir_data_size = le32_to_cpu(private->size) - SOF_EFFECT_PAYLOAD_OFFSET;
+	ipc_size = sizeof(struct sof_ipc_comp_eq_iir) + iir_data_size;
+	iir_data_p = private->data + SOF_EFFECT_PAYLOAD_OFFSET;
+
+	iir = kzalloc(ipc_size, GFP_KERNEL);
+	if (!iir)
+		return -ENOMEM;
+
+	/* configure iir IPC message */
+	iir->comp.hdr.size = ipc_size;
+	iir->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
+	iir->comp.id = swidget->comp_id;
+	iir->comp.type = SOF_COMP_EQ_IIR;
+	iir->comp.pipeline_id = index;
+
+	memcpy(&iir->data, iir_data_p, iir_data_size);
 
-	sof_dbg_comp_config(scomp, &eq->config);
+	iir->size = iir_data_size;
 
-	swidget->private = (void *)eq;
+	ret = sof_parse_tokens(scomp, &iir->config, comp_tokens,
+			       ARRAY_SIZE(comp_tokens), private->array,
+			       SOF_EFFECT_PAYLOAD_OFFSET);
```

Comment:
@juimonen I still this line should be le32_to_cpu(private->size) instead of SOF_EFFECT_PAYLOAD_OFFSET

Reply:
I think my previous comment was lost due to pr update. So sof_parse_tokens, assumes that the private->array consist solely of vendor tokens and it tries to parse them as much as   le32_to_cpu(private->size) is. Now if the private has also other stuff like bytes fields the function will try to read that as tokens (and will fail, because array size is random etc.). That's why I have to "limit" the size what is given to token parsing. Not sure then If I have understood the parsing incorrectly...?

Reply:
hmm ok. But I guess then we assume that the binary blob mixed with vendor arrays always contains first 4 bytes as size?

Reply:
@lgirdwood any comments how should I parse the bytes field mixed with vendor arrays in the widget private data? Also should I use the abi_hdr for sending the bytes to firmware?

Reply:
@juimonen iirc, this has it's own private data structure and does not use tuples. i.e. you create a private data structure and append the bytes to the end of it.

Reply:
@juimonen sorry it took me a while to get back to it. But I understand the sticky situation here. Can we pass this as a private data for the byte control associated with the EQ instead? Then there will not be any mixing of vendor tuples array and byte data.